compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,505 @@
+///   _1 = &raw? mut? _2;
+///   _3 = *_1;
+///   _4 = _1
+///   _5 = *_4
+/// In such cases, fully substituting `_1` means fully subtituting all of the copies.
+///
+/// For immutable borrows, we do not need to preserve such uniqueness property,
+/// so we perform all the possible substitutions without removing the `_1 = &_2` statement.
+///
+/// For simplicity, we treat raw pointers exactly like references.  This is sound because
+/// references impose strictly more constraints than raw pointers.
+pub struct PropagateReferences;
+
+impl<'tcx> MirPass<'tcx> for PropagateReferences {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        sess.mir_opt_level() >= 4
+    }
+
+    #[instrument(level = "trace", skip(self, tcx, body))]
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(def_id = ?body.source.def_id());
+        propagate_ssa(tcx, body);
+    }
+}
+
+fn propagate_ssa<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    let mut ssa = SsaLocals::new(body);
+
+    let copy_classes = ssa.compute_copy_classes(body);
+    debug!(?copy_classes);
+
+    let fully_replacable_locals = fully_replacable_locals(&ssa, &copy_classes);
+    debug!(?fully_replacable_locals);
+
+    let targets = compute_local_values(tcx, body, &ssa, &fully_replacable_locals);
+    debug!(?targets);
+
+    let mut any_replacement = false;
+    let mut storage_to_remove = BitSet::new_empty(body.local_decls.len());
+    for v in targets.iter() {
+        match v {
+            Value::Pointer(p) => {
+                if p.projection.first() == Some(&PlaceElem::Deref) {
+                    // We may introduce uses of a local after its `StorageDead`.
+                    // To avoid introducing UB, just remove the StorageDead statements.
+                    storage_to_remove.insert(p.local);
+                }
+                any_replacement = true;
+            }
+            Value::Unknown => {}
+        }
+    }
+
+    Replacer { tcx, targets, storage_to_remove }.visit_body_preserves_cfg(body);
+
+    if any_replacement {
+        crate::simplify::remove_unused_definitions(body);
+    }
+
+    verify_storage(tcx, body);
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum LocationExtended {
+    Plain(Location),
+    Arg,
+}
+
+#[derive(Debug)]
+struct SsaLocals {
+    /// Assignments to each local.  This defines whether the local is SSA.
+    assignments: IndexVec<Local, Set1<LocationExtended>>,
+    /// Number of "direct" uses of each local, ie. uses that are not dereferences.
+    /// We ignore non-uses (Storage statements, debuginfo).
+    direct_uses: IndexVec<Local, u32>,
+    /// We visit the body in reverse postorder, to ensure each local is assigned before it is used.
+    /// We remember the order in which we saw the assignments to compute the SSA values in a single
+    /// pass.
+    assignment_order: Vec<Local>,
+}
+
+impl SsaLocals {
+    fn new(body: &Body<'_>) -> SsaLocals {
```

Comment:
I'll test this later (no checkout on this machine), but does something like this ICE:

```rust
#![feature(core_intrinsics, custom_mir)]

use core::intrinsics::mir::*;

#[custom_mir(dialect = "runtime", phase = "post-cleanup")]
fn weird() {
    mir!({
        let y = x;
        let x = 5_i32;
        Return()
    })
}
```

More generally, can you explicitly document what the contract is that the locals that are "SSA locals" are expected to satisfy? Given that it says "SSA," I would assume "written to exactly once, never have an address/reference taken, and all uses dominate the assignment." But that's not what we're checking here so some clarification would help

Reply:
I missed the "all uses dominate the assignment" part. Is it an issue? If an use doesn't dominate the single assignment, we are reading from an uninitialized value, which is UB. So whether we perform the substitution doesn't matter.

Reply:
I'm for now not worried about correctness, but primarily about ICEs. I saw that you added the test, thanks for that. Why does it not ICE? The thing I'm worried about is that there are comments in places that like `copy_compute_classes` that expect SSA locals to only be used after they have shown up in the `assignment_order`

Reply:
If a use isn't dominated by the single assignment, it doesn't necessarily follow that the local is uninitialized. Using the other pull request:

```console
$ cat a.rs 
#![feature(custom_mir, core_intrinsics)]
#![allow(unused_assignments)]
extern crate core;
use core::intrinsics::mir::*;

#[custom_mir(dialect = "analysis", phase = "post-cleanup")]
fn f(c: bool) -> bool { mir!(
    let a: bool;
    let b: bool;
    { Goto(bb1) }
    bb1 = { b = c; match b { false => bb3, _ => bb2 }}
    bb2 = { a = b; c = false; Goto(bb1) }
    bb3 = { RET = a; Return() }
)}

fn main() {
    assert_eq!(true, f(true));
}
$ rustc +stage1 a.rs && ./a
$ rustc +stage1 a.rs -Zmir-enable-passes=+CopyProp && ./a
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `true`,
 right: `false`', a.rs:17:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
Thanks for the reproducer. Yeah, this is exactly what I was worried about